cdap-app-fabric/src/main/java/io/cdap/cdap/gateway/handlers/AppLifecycleHttpHandler.java

```diff
@@ -250,7 +260,52 @@ public void getAllApps(HttpRequest request, HttpResponder responder,
       .map(ApplicationRecord::new)
       .collect(Collectors.toList());
 
-    responder.sendJson(HttpResponseStatus.OK, GSON.toJson(apps));
+    if (Optional.ofNullable(pageSize).orElse(0) != 0) {
+      PaginatedApplicationRecords result = getPaginatedResults(apps, pageToken, pageSize, orderBy,
+                                                               nameFilter);
+      responder.sendJson(HttpResponseStatus.OK, GSON.toJson(result));
+    } else {
+      responder.sendJson(HttpResponseStatus.OK, GSON.toJson(apps));
+    }
+  }
+
+  /**
+   * Returns paginated result for getApps, also includes nextPageToken
+   *
+   * @param apps the applications obtained from the Store
+   * @param pageToken the next pageToken to return the paginated results from
+   * @param pageSize the total number of application records to display on the result page
+   * @param orderBy  ASC for ascending order, DESC for descending order
+   * @param nameFilter filter applications whose name contains nameFilter
+   */
+  private PaginatedApplicationRecords getPaginatedResults(List<ApplicationRecord> apps,
+                                                           String pageToken, Integer pageSize,
+                                                           SortOrder orderBy, String nameFilter
+      ) {
+
+    if (Strings.isNullOrEmpty(pageToken)) {
```

Comment:
It means first page, not last. Empty page token should mean index=0

Reply:
Missed that, will make the change.